src/app/modules/submission/email-submission/email-submission.controller.ts

```diff
@@ -30,267 +33,247 @@ const logger = createLoggerWithLabel(module)
+          )
+          .map((parsedResponses) => ({ parsedResponses, form }))
+          .mapErr(logErrorWithReqMeta('Error processing responses')),
+      )
+      .andThen(({ parsedResponses, form }) => {
+        const { authType } = form
+        switch (authType) {
+          case AuthType.CP:
+            return SpcpFactory.extractJwt(req.cookies, authType)
+              .asyncAndThen((jwt) => SpcpFactory.extractCorppassJwtPayload(jwt))
+              .map((jwt) => ({
+                form,
+                parsedResponses: [
+                  ...parsedResponses,
+                  ...createCorppassParsedResponses(jwt.userName, jwt.userInfo),
+                ],
+                hashedFields: new Set<string>(),
+              }))
+              .mapErr((error) => {
+                spcpSubmissionFailure = true
+                return logErrorWithReqMeta(
+                  'Failed to verify JWT with auth client',
+                )(error)
+              })
+          case AuthType.SP:
+            return SpcpFactory.extractJwt(req.cookies, authType)
+              .asyncAndThen((jwt) => SpcpFactory.extractSingpassJwtPayload(jwt))
+              .map((jwt) => ({
+                form,
+                parsedResponses: [
+                  ...parsedResponses,
+                  ...createSingpassParsedResponses(jwt.userName),
+                ],
+                hashedFields: new Set<string>(),
+              }))
+              .mapErr((error) => {
+                spcpSubmissionFailure = true
+                return logErrorWithReqMeta(
+                  'Failed to verify JWT with auth client',
+                )(error)
+              })
+          case AuthType.MyInfo:
+            return MyInfoUtil.extractMyInfoCookie(req.cookies)
+              .andThen(MyInfoUtil.extractAccessTokenFromCookie)
+              .andThen((accessToken) =>
+                MyInfoFactory.extractUinFin(accessToken),
+              )
+              .asyncAndThen((uinFin) =>
+                MyInfoFactory.fetchMyInfoHashes(uinFin, formId)
+                  .andThen((hashes) =>
+                    MyInfoFactory.checkMyInfoHashes(parsedResponses, hashes),
+                  )
+                  .map((hashedFields) => ({
+                    form,
+                    hashedFields,
+                    parsedResponses: [
+                      ...parsedResponses,
+                      ...createSingpassParsedResponses(uinFin),
+                    ],
+                  })),
+              )
+              .mapErr((error) => {
+                spcpSubmissionFailure = true
+                return logErrorWithReqMeta('Error verifying MyInfo hashes')(
+                  error,
+                )
+              })
+          default:
+            return ok({
+              form,
+              parsedResponses,
+              hashedFields: new Set<string>(),
+            }) as Result<
+              {
```

Comment:
i agree with the first point. extracting it has value, because it helps the reader to understand this whole chunk without getting lost in the details.

for the second point, this is to aid in typehinting for typescript, and even if we were to add it as a generic argument to this particular `andThen`, we would still need the annotation on the errors just so that type level information is preserved and not lost (because of the `ok`).

i think this would impact readability just by virtue of how verbose it is (due to the errors) and wouldn't significantly impact clarity because the destructuring of the next `andThen` makes the shape and types clear